framer/Context.coffee

```diff
@@ -1,122 +1,241 @@
-		@_animationList = []
-		@_delayTimers = []
-		@_delayIntervals = []
-		@_layerIdCounter = 1
+		@resetLayers()
+		@resetAnimations()
+		@resetTimers()
+		@resetIntervals()
 
 		@emit("reset", @)
 
-	destroy: ->
-		@reset()
-		if @_rootElement.parentNode
-			@_rootElement.parentNode.removeChild(@_rootElement)
-		Utils.domCompleteCancel(@_appendRootElement)
+	# destroy: ->
+	# 	@reset()
 
-	getRootElement: ->
-		@_rootElement
 
-	getLayers: ->
-		_.clone(@_layerList)
+	##############################################################
+	# Collections
 
+	# Layers
+	@define "layers", get: -> _.clone(@_layers)
+	@define "layerCounter", get: -> @_layerCounter
+	
 	addLayer: (layer) ->
-		return if layer in @_layerList
-		@_layerList.push(layer)
-		return null
-
+		return if layer in @_layers
+		@_layerCounter++
+		@_layers.push(layer)
+		
 	removeLayer: (layer) ->
-		@_layerList = _.without(@_layerList, layer)
-		return null
+		@_layers = _.without(@_layers, layer)
+	
+	resetLayers: ->
+		@_layers = []
+		@_layerCounter = 0
 
-	layerCount: ->
-		return @_layerList.length
 
-	nextLayerId: ->
-		@_layerIdCounter++
+	# Animations
+	@define "animations", get: -> _.clone(@_animations)
+	
+	addAnimation: (animation) ->
+		return if animation in @_animations
+		@_animations.push(animation)
+		
+	removeAnimation: (animation) ->
+		@_animations = _.without(@_animations, animation)
+	
+	resetAnimations: ->
+		@stopAnimations()
+		@_animations = []
 
-	_createRootElement: ->
+	stopAnimations: ->
+		return unless @_animations
+		@_animations.map (animation) -> animation.stop(true)
 
-		@_rootElement = document.createElement("div")
-		@_rootElement.id = "FramerContextRoot-#{@_name}"
-		@_rootElement.classList.add("framerContext")
 
-		if @_parentLayer
-			@_appendRootElement()
-		else
-			Utils.domComplete(@_appendRootElement)
+	# Timers
+	@define "timers", get: -> _.clone(@_timers)
+	
+	addTimer: (timer) ->
+		return if timer in @_timers
+		@_timers.push(timer)
+		
+	removeTimer: (timer) ->
+		@_timers = _.without(@_timers, timer)
+	
+	resetTimers: ->
+		@_timers.map(window.clearTimeout) if @_timers
+		@_timers = []
 
-	_appendRootElement: =>
-		parentElement = @_parentLayer?._element
-		parentElement ?= document.body
-		parentElement.appendChild(@_rootElement)
 
-	run: (f) ->
+	# Intervals
+	@define "intervals", get: -> _.clone(@_intervals)
+	
+	addInterval: (interval) ->
+		return if interval in @_intervals
+		@_intervals.push(interval)
+		
+	removeInterval: (interval) ->
+		@_intervals = _.without(@_intervals, interval)
+	
+	resetIntervals: ->
+		@_intervals.map(window.clearInterval) if @_intervals
+		@_intervals = []
+
+
+	##############################################################
+	# Run
+
+	run: (fn) ->
 		previousContext = Framer.CurrentContext
 		Framer.CurrentContext = @
-		f()
+		fn()
 		Framer.CurrentContext = previousContext
 
+
+	##############################################################
+	# Freezing
+
+	freeze: ->
+
+		if @_frozenEvents
```

Comment:
`if @_frozenEvents? ` is syntactically nicer here, 'cause it will return `true` when `@_frozenEvents` is `null` or `undefined`. Not that it matters functionally, in this case though :-)